src/EditorFeatures/Core.Wpf/NavigateTo/NavigateToItemProvider.Callback.cs

```diff
@@ -0,0 +1,140 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.CodeAnalysis.NavigateTo;
+using Microsoft.CodeAnalysis.Text.Shared.Extensions;
+using Microsoft.VisualStudio.Language.NavigateTo.Interfaces;
+using Microsoft.VisualStudio.Text.PatternMatching;
+using Roslyn.Utilities;
+
+namespace Microsoft.CodeAnalysis.Editor.Implementation.NavigateTo
+{
+    internal partial class NavigateToItemProvider
+    {
+        private class VisualStudioNavigateToCallback : INavigateToSearchCallback
+        {
+            private readonly INavigateToItemDisplayFactory _displayFactory;
+            private readonly INavigateToCallback _callback;
+
+            public VisualStudioNavigateToCallback(INavigateToItemDisplayFactory displayFactory, INavigateToCallback callback)
+            {
+                _displayFactory = displayFactory;
+                _callback = callback;
+            }
+
+            public void Done(bool isFullyLoaded = true)
+            {
+                if (!isFullyLoaded && _callback is INavigateToCallback2 callback2)
+                {
+                    callback2.Done(IncompleteReason.SolutionLoading);
+                }
+                else
+                {
+                    _callback.Done();
+                }
+            }
+
+            public Task AddItemAsync(Project project, INavigateToSearchResult result, CancellationToken cancellationToken)
+            {
+                ReportMatchResult(project, result);
+                return Task.CompletedTask;
+            }
+
+            public void ReportProgress(int current, int maximum)
+            {
+                _callback.ReportProgress(current, maximum);
+            }
+
+            private void ReportMatchResult(Project project, INavigateToSearchResult result)
```

Comment:
here and below is just copied exactly from https://github.com/dotnet/roslyn/blob/master/src/EditorFeatures/Core.Wpf/NavigateTo/NavigateToItemProvider.Searcher.cs#L216